### PR TITLE
gemm + layernorm

### PR DESCRIPTION
**Pull request by rocking5566**
Hardcode the 5ary-elementwise op now.
Implement generic elementwse op in the future.

### CODE REVIEW

File: example/16_gemm_reduce/gemm_reduce_xdl_mean_squaremean_fp16.cpp [hunk 1]
```diff
@@ -47,10 +47,12 @@ using DxsReduceOp = ck::Tuple<D0ReduceOp, D1ReduceOp>;
 
 using UnaryIdenticElementOp =
     ck::tensor_operation::element_wise::UnaryIdentic<ReduceAccDataType, ReduceAccDataType, false>;
+using UnaryDivElementOp =
```
Comment (asroy) [inline example/16_gemm_reduce/gemm_reduce_xdl_mean_squaremean_fp16.cpp:50]: @qianfengz

It is so strange to use ```UnaryIdentic``` for division.

The pointwise function used by reduction operator should be refactored to use the same pointwise functions as other operations. Please produce a plan for this change

File: include/ck/tensor_operation/gpu/grid/gridwise_5ary_Elementwise_1d.hpp [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+          typename DDataType,
+          typename EDataType,
+          typename FDataType,
+          typename AGridDesc_M,
+          typename BGridDesc_M,
+          typename CGridDesc_M,
+          typename DGridDesc_M,
+          typename EGridDesc_M,
+          typename FGridDesc_M,
+          typename ElementwiseFunctor>
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/grid/gridwise_5ary_Elementwise_1d.hpp:23]: A/B/C/D/E should all have their own ```ScalarPerVector```, not using vector access means the performance would be bad

File: example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+using UnaryIdenticElementOp =
+    ck::tensor_operation::element_wise::UnaryIdentic<ReduceAccDataType, ReduceAccDataType, false>;
+using UnaryDivElementOp =
+    ck::tensor_operation::element_wise::UnaryIdentic<ReduceAccDataType, ReduceAccDataType, true>;
+using UnarySquareElementOp =
+    ck::tensor_operation::element_wise::UnarySquare<ReduceAccDataType, ReduceAccDataType, false>;
+using DxsInElementOp  = ck::Tuple<UnaryIdenticElementOp, UnarySquareElementOp>;
+using DxsOutElementOp = ck::Tuple<UnaryDivElementOp, UnaryDivElementOp>;
+
+using DGlobalMemOp =
```
Comment (asroy) [inline example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp:60]: DxsGlobalMemOp

File: example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+        throw std::runtime_error(
+            "wrong! device_gemm with the specified compilation parameters does "
+            "not support this GEMM problem");
+    }
+
+    reduceMean_device_buf.SetZero();
+    reduceMeanSquare_device_buf.SetZero();
+
+    // Prepare LayerNorm
+    auto layerNorm             = DeviceNormalizeInstance{};
```
Comment (asroy) [inline example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp:277]: This is not "layer norm". It's normalization phase of "layer nomr"

File: example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    auto layerNorm_invoker_ptr = layerNorm.MakeInvokerPointer();
```
Comment (asroy) [inline example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp:278]: gemmReduce is not pointer, but layerNorm_invoker_ptr is a pointer. Please be consistent

File: example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp [hunk 6]
```diff
@@ -0,0 +1,331 @@
+#include <iostream>
```
Comment (asroy) [inline example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp:1]: this folder name should be ```21_gemm_layernorm```

**Review by asroy — CHANGES_REQUESTED** [review]
some cosmetic changes, otherwise LGTM

File: example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+    if(!layerNorm.IsSupportedArgument(layerNorm_argument.get()))
+    {
+        throw std::runtime_error("The runtime parameters seems not supported by the "
+                                 "Device5AryElementwise_Xdl_CShuffle instance, exiting!");
+    }
+
+    // run kernel
+    gemmReduce_invoker.Run(gemmReduce_argument, StreamConfig{nullptr, time_kernel});
+    layerNorm_invoker_ptr->Run(layerNorm_argument.get(), StreamConfig{nullptr, time_kernel});
+
```
Comment (asroy) [inline example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp:304]: Need to output Performance information:
1. for GEMM+reduce: kernel time, tflops (use fma flops only, no need to calculte reduction flops), bandwidth
2. for pintwise: kernel time, and bandwidth

File: example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp (hunk 4 above)
Comment (rocking5566) [reply example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp:277]: I do not use normalize any more. Use 5-nary elementwise instead

File: example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp (hunk 3 above)
Comment (rocking5566) [reply example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp]: 2fc2a1893fc6d7bc0e226829f019b2dd4886a22d

File: example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp (hunk 4 above)
Comment (rocking5566) [reply example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp]: 2fc2a1893fc6d7bc0e226829f019b2dd4886a22d

File: example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp (hunk 5 above)
Comment (rocking5566) [reply example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp]: da8c060873fb9120004627b2d3963154e1f8a68c

File: example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp (hunk 6 above)
Comment (rocking5566) [reply example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp]: 2bd585e4fb05f581295681dff0684fd8c18f748b

File: include/ck/tensor_operation/gpu/device/device_5ary_elementwise_xdl_cshuffle.hpp [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+          typename ElementwiseFunctor,
+          index_t NDim,
+          index_t MPerThread,
+          index_t AScalarPerVector,
+          index_t BScalarPerVector,
+          index_t CScalarPerVector,
+          index_t DScalarPerVector,
+          index_t EScalarPerVector,
+          index_t FScalarPerVector>
+struct Device5AryElementwise_Xdl_CShuffle : public BaseOperator
```
[the remark this answers is not in the document]
Comment (rocking5566) [reply include/ck/tensor_operation/gpu/device/device_5ary_elementwise_xdl_cshuffle.hpp:32]: 9402ee4bae4ee0c7be74206b1d8a4e472fd0103f

[the remark this answers is not in the document]
Comment (rocking5566) [reply include/ck/tensor_operation/gpu/device/device_5ary_elementwise_xdl_cshuffle.hpp]: 9402ee4bae4ee0c7be74206b1d8a4e472fd0103f

File: include/ck/tensor_operation/gpu/device/device_normalize_xdl_cshuffle.hpp [hunk 9]
```diff
@@ -83,7 +83,6 @@ struct DeviceNormalize_Xdl_CShuffle : public BaseOperator
                  const GammaDataType* p_gamma,
                  const BetaDataType* p_beta,
                  OutDataType* p_output,
-                 MeanSquareDataType epsilon,
```
[the remark this answers is not in the document]
Comment (rocking5566) [reply include/ck/tensor_operation/gpu/device/device_normalize_xdl_cshuffle.hpp]: I used 5-nary elementwise instead. This op should not know the epsilon.
I move it to the lambda

File: example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+    }
+
+    reduceMean_device_buf.SetZero();
+    reduceMeanSquare_device_buf.SetZero();
+
+    // Prepare LayerNorm
+    auto layerNorm             = DeviceNormalizeInstance{};
+    auto layerNorm_invoker_ptr = layerNorm.MakeInvokerPointer();
+    auto layerNorm_argument =
+        layerNorm.MakeArgumentPointer(c_device_buf.GetDeviceBuffer(),
```
[the remark this answers is not in the document]
Comment (rocking5566) [reply example/21_gemm_normalize_xdl/gemm_layernorm_xdl_fp16.cpp]: I implement gemm + reduction + normalize(x, E[x], E[x^2], gamma, beta)
gemm + bias is wait until next PR

**Review by asroy — APPROVED** [review]

**Pull request merged**